### PR TITLE
PV Smoothing to merge into develop to consolidate FOM options

### PR DESCRIPTION
**Pull request by sjanzou**

### CODE REVIEW

File: deploy/runtime/ui/Battery Dispatch PV Smoothing.txt
Comment (mjprilliman) [inline deploy/runtime/ui/Battery Dispatch PV Smoothing.txt:1869]: Change toggle_dispatch to toggle_dispatch_pvs as Manual Dispatch also uses toggle_dispatch()

File: deploy/runtime/ui/Battery Dispatch Automated FOM.txt
Comment (mjprilliman) [inline deploy/runtime/ui/Battery Dispatch Automated FOM.txt:37]: Index Labels property on variables was overwriting Object properties Items list so PV Smoothing wasn't showing up. I committed the fix to this branch

File: test_results_win64.csv
Comment (mjprilliman) [inline test_results_win64.csv:127]: Need to re-run Single Owner model as the defaults were all changed in the PVS debugging process.

**Comment by brtietz** [comment]
Given that the weather file default in SAM is an hour I wonder if this red message should explicitly discuss a shorter timestep weather file:

image

**Review by mjprilliman — COMMENTED** [review]
Fixed some issues from the integration of develop into PVSmoothing.

**Comment by brtietz** [comment]
This is the error I get when trying to select PVSmoothing on the Fuel Cell - Single Owner configuration:

image

Do we want to provide the option for that configuration or remove it?

**Comment by brtietz** [comment]
Looks like there are a few issues with generic battery FOM cases as well. This is with "import hourly or subhourly generation profile" selected:

image

**Review by brtietz — CHANGES_REQUESTED** [review]
Detailed PV-Battery FOM configurations look good to me. Fuel Cell and Generic System cases have some bugs with the new PV Smoothing UI.

**Comment by mjprilliman** [comment]
> This is the error I get when trying to select PVSmoothing on the Fuel Cell - Single Owner configuration:
> 
> image
> 
> Do we want to provide the option for that configuration or remove it?

Not sure on whether it should be included or not but this was a startup.lk error as the PV Smoothing form wasn't in the list for Fuel Cell FOM. I can push a fix if we do want to include this as an option

**Comment by mjprilliman** [comment]
> Looks like there are a few issues with generic battery FOM cases as well. This is with "import hourly or subhourly generation profile" selected:
> 
> image

This stems from the ui_step_minutes calculated variable. The issue there is that when you select "import hourly or subhourly generation profile", The default for the energy_output_array variable has size 1. So is this a bug or a hint to the user that they actually need to input some sort of generation profile?

**Review by cpaulgilman — APPROVED** [review]
I did a little UI cleanup, including addressing @brtietz's comment about the warning message for subhourly weather data:

image

I think this option should only be available for PV Battery / Single Owner configurations for now. I could see an application for Generic Battery / Single Owner for users who want to run the model with measured output data, but this mostly seems to be a forecasting tool rather than a performance monitoring tool. For fuel cells would people expect an option to allow the fuel cell to contribute to PV ramp rate mitigation?

I pushed a change to handle that for now by displaying a UI message on Battery Dispatch Options FOM if `batt_dispatch_choice_ui = 3` when technology() != PV Battery, and come up with a better solution later.

**Comment by cpaulgilman** [comment]
> 
> 
> > This is the error I get when trying to select PVSmoothing on the Fuel Cell - Single Owner configuration:
> > image
> > Do we want to provide the option for that configuration or remove it?
> 
> Not sure on whether it should be included or not but this was a startup.lk error as the PV Smoothing form wasn't in the list for Fuel Cell FOM. I can push a fix if we do want to include this as an option

@mjprilliman Can you push a fix to add the PV smoothing UI form to Fuel Cell configurations for now? After my latest commit, if you choose PV smoothing a red warning message should display saying that PV smoothing is for PV Battery only. Otherwise we will need a different version of batt_dispatch_choice_ui for PV and non-PV battery configs.

**Comment by mjprilliman** [comment]
> > > This is the error I get when trying to select PVSmoothing on the Fuel Cell - Single Owner configuration:
> > > image
> > > Do we want to provide the option for that configuration or remove it?
> > 
> > 
> > Not sure on whether it should be included or not but this was a startup.lk error as the PV Smoothing form wasn't in the list for Fuel Cell FOM. I can push a fix if we do want to include this as an option
> 
> @mjprilliman Can you push a fix to add the PV smoothing UI form to Fuel Cell configurations for now? After my latest commit, if you choose PV smoothing a red warning message should display saying that PV smoothing is for PV Battery only. Otherwise we will need a different version of batt_dispatch_choice_ui for PV and non-PV battery configs.

@cpaulgilman This wouldn't exactly be an elegant solution but could we change the Height property on batt_dispatch_choice_ui for non PV-Battery configs so that users would never see PV Smoothing? It will not be the default for any cases so the PV Smoothing UI form would only load in the case of someone changing the value of batt_dispatch_choice_ui from the Inputs Browser, which would then be covered by your message.

**Comment by cpaulgilman** [comment]
> 
> 
> > > > This is the error I get when trying to select PVSmoothing on the Fuel Cell - Single Owner configuration:
> > > > image
> > > > Do we want to provide the option for that configuration or remove it?
> > > 
> > > 
> > > Not sure on whether it should be included or not but this was a startup.lk error as the PV Smoothing form wasn't in the list for Fuel Cell FOM. I can push a fix if we do want to include this as an option
> > 
> > 
> > @mjprilliman Can you push a fix to add the PV smoothing UI form to Fuel Cell configurations for now? After my latest commit, if you choose PV smoothing a red warning message should display saying that PV smoothing is for PV Battery only. Otherwise we will need a different version of batt_dispatch_choice_ui for PV and non-PV battery configs.
> 
> @cpaulgilman This wouldn't exactly be an elegant solution but could we change the Height property on batt_dispatch_choice_ui for non PV-Battery configs so that users would never see PV Smoothing? It will not be the default for any cases so the PV Smoothing UI form would only load in the case of someone changing the value of batt_dispatch_choice_ui from the Inputs Browser, which would then be covered by your message.

Thinking outside of the box, eh? ;)

I'm not sure that will work across operating systems because of the difference in font size. I think you could use the property() function to conditionally set the Items property of the batt_dispatch_choice widget, but then that would not be reflected in the Index Labels for that variable, so the option would appear in Parametrics and Inputs Browser. It's also a hack that might cause confusion for someone besides us down the road trying to understand how the form works..

**Comment by mjprilliman** [comment]
> > > > > This is the error I get when trying to select PVSmoothing on the Fuel Cell - Single Owner configuration:
> > > > > image
> > > > > Do we want to provide the option for that configuration or remove it?
> > > > 
> > > > 
> > > > Not sure on whether it should be included or not but this was a startup.lk error as the PV Smoothing form wasn't in the list for Fuel Cell FOM. I can push a fix if we do want to include this as an option
> > > 
> > > 
> > > @mjprilliman Can you push a fix to add the PV smoothing UI form to Fuel Cell configurations for now? After my latest commit, if you choose PV smoothing a red warning message should display saying that PV smoothing is for PV Battery only. Otherwise we will need a different version of batt_dispatch_choice_ui for PV and non-PV battery configs.
> > 
> > 
> > @cpaulgilman This wouldn't exactly be an elegant solution but could we change the Height property on batt_dispatch_choice_ui for non PV-Battery configs so that users would never see PV Smoothing? It will not be the default for any cases so the PV Smoothing UI form would only load in the case of someone changing the value of batt_dispatch_choice_ui from the Inputs Browser, which would then be covered by your message.
> 
> Thinking outside of the box, eh? ;)
> 
> I'm not sure that will work across operating systems because of the difference in font size. I think you could use the property() function to conditionally set the Items property of the batt_dispatch_choice widget, but then that would not be reflected in the Index Labels for that variable, so the option would appear in Parametrics and Inputs Browser. It's also a hack that might cause confusion for someone besides us down the road trying to understand how the form works..

@cpaulgilman Trying this with the Height property did not update the widget on the first load for whatever reason. Changing the Items property does work but I ran into the same confusion you mention. I think I will just rely on your message and leave it at that. Should I disable the PV Smoothing inputs for non PV Battery configs? Right now SAM will still run for Generic Battery or Fuel Cell and the PV Smoothing option selected, the results probably just don't make much sense

**Comment by sjanzou** [comment]
I think we should disable PV Smoothing for all but detailed PV. We discussed that smoothing is only an option for PV systems. I see where smoothing may be useful for the PV Fuel Cell combo and i see some use cases for generic.

The red error message was directly from Will. Smoothing does not make sense at weather file intervals greater than 15 minutes per the EPRI meetings.

Let me know what you think and i will update the pull request.

Thanks for the detailed reviews!

**Comment by mjprilliman** [comment]
@sjanzou @cpaulgilman Take a look at the latest commit and see if that is sufficient or if we need to more on non PV Battery

**Comment by cpaulgilman** [comment]
> 
> 
> @sjanzou @cpaulgilman Take a look at the latest commit and see if that is sufficient or if we need to more on non PV Battery

This looks good to me. I revised the message to mention 15 minute weather data instead of 30 minutes.

**Comment by sjanzou** [comment]
Looks good to me. Any other items for this pull request?

**Comment by mjprilliman** [comment]
> Looks good to me. Any other items for this pull request?

Nothing from me

**Comment by cpaulgilman** [comment]
> 
> 
> > Looks good to me. Any other items for this pull request?
> 
> Nothing from me

Nor from me.

**Comment by sjanzou** [comment]
Generic Battery and Fuel Cell updates for FOM dispatch with PV Smoothing updated and ssc tests added and project files tested. Please recheck and let me know if you have any changes. Thanks!

File: deploy/runtime/startup.lk
Comment (mjprilliman) [inline deploy/runtime/startup.lk:562]: So @cpaulgilman and I have discussed this issue. The battery manual dispatch page for fuel cell also includes the controls for fuel cell manual dispatch. So if the user selects Manual Dispatch for fuel cell dispatch but not for battery dispatch, they will not be able to see the controls for manual fuel cell dispatch. This was why the Fuel Cell Dispatch Manual was fixed as a header page.

File: deploy/runtime/ui/Battery Dispatch Options FOM.txt
Comment (mjprilliman) [inline deploy/runtime/ui/Battery Dispatch Options FOM.txt]: So is the final decision to give a warning to the user that the PVSmoothing option is only intended for PV Battery configs or not?

File: deploy/runtime/ui/Battery Dispatch PV Smoothing.txt
Comment (mjprilliman) [inline deploy/runtime/ui/Battery Dispatch PV Smoothing.txt:1775]: If we are hard-coding this then `mode` is not needed.

**Review by mjprilliman — COMMENTED** [review]
A little confused now and what the final verdict is on the Fuel Cell FOM handling

File: test_results_win64.csv
Comment (mjprilliman) [inline test_results_win64.csv:124]: These are the test values that are in Patch 2. I changed the values with this commit (https://github.com/NREL/SAM/commit/6b8ebd84e106c7b5158b71d49bddbf984fbaa4c9#diff-8b237736d5563a32f81df3035be5cccb0f7b2cfd8b863b1d3d4e5956f65dfb1dR123) and must not have changed it back.

**Review by brtietz — APPROVED** [review]
Changes look good to me. Ok to merge, contingent on SSC issues being solved. I'm fine with resolving the manual dispatch UI form issues in another PR.

**Pull request merged**
